internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -0,0 +1,117 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package customresource provides functionality to upload Copilot custom resources.
+package asset
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+)
+
+type filterAction int
+
+const (
+	include filterAction = iota + 1
+	exclude
+)
+
+// UploadFunc is the function signature to upload contents to a destination.
+type UploadFunc func(dest string, contents io.Reader) (url string, err error)
+
+// UploadInput is the input of Upload.
+type UploadInput struct {
+	Source      string
+	Destination string
+	Includes    []string
+	Excludes    []string
+	Upload      UploadFunc
+	Reader      *afero.Afero
+}
+
+// Upload uploads static assets to Cloud Storage.
+func Upload(in *UploadInput) ([]string, error) {
```

Comment:
one pattern that I find kinda neat is separating _required_ parameters from _optional_ ones. For example, here it would be:
```go
type UploadOpts struct {
   Includes []string // Relative path under source to include in the upload.
   Excludes []string // Relative path under source to include in the upload.
   UploadFn UploadFunc // Custom implementation on how to upload the contents under a file. Defaults to S3UploadFn.
}

func Upload(fs afero.Fs, source, destination string, opts UploadOpts)
```

What do you think boss? We do something similar in `override/cdk.go`